File: 00.data-structure/cpp/leetcode_33.cpp
#include <iostream>

using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int pivot = findMinimumIndex(nums);

        if (pivot == 0) 
            return binarySearch(nums, target, 0, nums.size() - 1);

        if (target >= nums[0])
            return binarySearch(nums, target, 0, pivot - 1);
        else {
            return binarySearch(nums, target, pivot, nums.size() - 1);
        }
    }

    int findMinimumIndex(vector<int>& nums) {
        int left = 0;
        int right = nums.size() - 1;

        while (left < right) {
            if (nums[left] < nums[right]) 
                break;

            int mid = (left + right) / 2;

            if (nums[mid] > nums[right]) {
                left = mid + 1;
            } 
            else {
                right = mid;
            }
        }

        return left;
    }

    // arr 에서 조회 범위를 인자로 전달 받아서 추가 메모리 할당을 없앨 수 있음
    int binarySearch(vector<int>& nums, int target, int start, int end) {
        int left = start;
        int right = end;

        while (left <= right) {
            int mid = (left + right) / 2;

            if (nums[mid] == target) 
                return mid;
            else if (nums[mid] > target) 
                right = mid - 1;
            else 
                left = mid + 1;
        }

        return -1;
    }
};

int main() {
    Solution solution;

    vector<int> case1 = {4, 5, 6, 7, 0, 1, 2};
    cout << "input : {4, 5, 6, 7, 0, 1, 2}, target : 0 => answer : " << solution.search(case1, 0) << endl;
    cout << "input : {4, 5, 6, 7, 0, 1, 2}, target : 4 => answer : " << solution.search(case1, 4) << endl;
    cout << "input : {4, 5, 6, 7, 0, 1, 2}, target : 3 => answer : " << solution.search(case1, 3) << endl;

    vector<int> case2 = {1};
    cout << "input : {1}, target : 0 => answer : " << solution.search(case2, 0) << endl;
    cout << "input : {1}, target : 1 => answer : " << solution.search(case2, 1) << endl;

    vector<int> case3 = {1, 3};
    cout << "input : {1, 3}, target : 1 => answer : " << solution.search(case3, 1) << endl;

    return 0;
}

File: 00.data-structure/python/leetcode_33.py
"""
search in rotated sorted array
"""
def search(nums, target):
    # find k
    left = 0
    right = len(nums) - 1

    while left < right:
        if nums[left] < nums[right]:
            break

        mid = (left + right) // 2
        if nums[mid] > nums[right]:
            left = mid + 1
        else:
            # right = mid - 1
            right = mid  # why ??

    def find(arr, t):
        start = 0
        end = len(arr) - 1

        while start <= end:
            mid = (start + end) // 2

            if arr[mid] == t:
                return mid
            elif arr[mid] > t:
                end = mid - 1
            else:
                start = mid + 1

        return -1

    # find target in left
    # if left == 0, then not rotated
    if left == 0:
        return find(nums, target)

    if target >= nums[0]:
        return find(nums[:left], target)
    else:
        ans = find(nums[left:], target)
        if ans != -1:
            return ans + left
        else:
            return ans


search([4,5,6,7,0,1,2], 0)
search([4,5,6,7,0,1,2], 3)
search([4,5,6,7,0,1,2], 2)
search([5,6,7,0,1,2,4], 5)
search([1], 0)
search([1], 1)
search([1, 3], 1)  # invalid test case with the condition k >= 1
